File: EMS-MiniProject2/EMS.API/Services/AuthService.cs
using System.IdentityModel.Tokens.Jwt;
using System.Security.Claims;
using System.Text;
using EMS.API.Data;
using EMS.API.DTOs;
using EMS.API.Models;
using Microsoft.EntityFrameworkCore;
using Microsoft.IdentityModel.Tokens;

namespace EMS.API.Services
{
    public class AuthService
    {
        private readonly AppDbContext _db;
        private readonly IConfiguration _config;

        public AuthService(AppDbContext db, IConfiguration config)
        {
            _db     = db;
            _config = config;
        }

        public async Task<AuthResponseDto> RegisterAsync(RegisterRequestDto request)
        {
            // Validate role
            var role = request.Role?.Trim();
            if (role != "Admin" && role != "Viewer")
                role = "Viewer";

            // Check duplicate username (case-insensitive)
            var exists = await _db.Users
                .AnyAsync(u => u.Username.ToLower() == request.Username.Trim().ToLower());
            if (exists)
                return new AuthResponseDto { Success = false, Message = "Username already taken. Please choose another." };

            var user = new AppUser
            {
                Username     = request.Username.Trim(),
                PasswordHash = BCrypt.Net.BCrypt.HashPassword(request.Password, 12),
                Role         = role,
                CreatedAt    = DateTime.UtcNow,
            };

            _db.Users.Add(user);
            await _db.SaveChangesAsync();

            return new AuthResponseDto
            {
                Success  = true,
                Username = user.Username,
                Role     = user.Role,
                Message  = "Registration successful.",
            };
        }

        public async Task<AuthResponseDto> LoginAsync(LoginRequestDto request)
        {
            var user = await _db.Users
                .FirstOrDefaultAsync(u => u.Username.ToLower() == request.Username.Trim().ToLower());

            if (user == null || !BCrypt.Net.BCrypt.Verify(request.Password, user.PasswordHash))
                return new AuthResponseDto { Success = false, Message = "Invalid credentials. Please try again." };

            var token = GenerateToken(user);

            return new AuthResponseDto
            {
                Success  = true,
                Username = user.Username,
                Role     = user.Role,
                Token    = token,
                Message  = "Login successful.",
            };
        }

        private string GenerateToken(AppUser user)
        {
            var claims = new[]
            {
                new Claim(ClaimTypes.NameIdentifier, user.Id.ToString()),
                new Claim(ClaimTypes.Name,           user.Username),
                new Claim(ClaimTypes.Role,           user.Role),
            };

            var key   = new SymmetricSecurityKey(Encoding.UTF8.GetBytes(_config["Jwt:Key"]!));
            var creds = new SigningCredentials(key, SecurityAlgorithms.HmacSha256);

            var token = new JwtSecurityToken(
                issuer:             _config["Jwt:Issuer"],
                audience:           _config["Jwt:Audience"],
                claims:             claims,
                expires:            DateTime.UtcNow.AddHours(double.Parse(_config["Jwt:ExpiryHours"] ?? "8")),
                signingCredentials: creds
            );

            return new JwtSecurityTokenHandler().WriteToken(token);
        }
    }
}

File: EMS-MiniProject1/js/authService.js
const authService = (function () {
  let _registeredUsers = [{ ...INITIAL_ADMIN }];
  let _isLoggedIn = false;
  let _currentUser = null;

  return {
    signup(username, password) {
      const errors = {};
      if (!username || username.trim() === '') errors.username = 'Username is required.';
      if (!password || password.length < 6)    errors.password = 'Password must be at least 6 characters.';
      if (Object.keys(errors).length) return { success: false, errors };

      const exists = _registeredUsers.find(
        u => u.username.toLowerCase() === username.trim().toLowerCase()
      );
      if (exists) return { success: false, errors: { username: 'Username already taken. Please choose another.' } };

      _registeredUsers.push({ username: username.trim(), password });
      return { success: true };
    },

    login(username, password) {
      if (!username || !password) return { success: false, error: 'Both fields are required.' };
      const user = _registeredUsers.find(u => u.username === username.trim() && u.password === password);
      if (!user) return { success: false, error: 'Invalid credentials. Please try again.' };
      _isLoggedIn = true;
      _currentUser = user.username;
      return { success: true };
    },

    logout() {
      _isLoggedIn = false;
      _currentUser = null;
    },

    isLoggedIn()     { return _isLoggedIn; },
    getCurrentUser() { return _currentUser; },

    _reset() {
      _registeredUsers = [{ ...INITIAL_ADMIN }];
      _isLoggedIn = false;
      _currentUser = null;
    },
  };
})();

if (typeof module !== 'undefined' && module.exports) {
  module.exports = { authService };
}
